core/src/plugins/kubernetes/results-cache-base.ts

```diff
@@ -0,0 +1,207 @@
+import { runResultSchemaZod } from "../../plugin/base.js"
+import type { NamespaceStatus } from "../../types/namespace.js"
+import { namespaceStatusSchema } from "../../types/namespace.js"
+import type { RunAction } from "../../actions/run.js"
+import type { TestAction } from "../../actions/test.js"
+import { hashSync } from "hasha"
+import { Memoize } from "typescript-memoize"
+import type { AnyZodObject, z } from "zod"
+import { deline } from "../../util/string.js"
+import type { ContainerRunAction, ContainerTestAction } from "../container/config.js"
+import type { KubernetesRunAction, KubernetesTestAction } from "./kubernetes-type/config.js"
+import type { HelmPodRunAction, HelmPodTestAction } from "./helm/config.js"
+import { renderZodError } from "../../config/zod.js"
+import type { JsonObject } from "type-fest"
+import { GardenError } from "../../exceptions.js"
+
+export type CacheableAction = RunAction | TestAction
+
+export type CacheableRunAction = ContainerRunAction | KubernetesRunAction | HelmPodRunAction
+export type CacheableTestAction = ContainerTestAction | KubernetesTestAction | HelmPodTestAction
+
+export type SchemaVersion = `v${number}`
+
+/**
+ * Increment the current result schema format version
+ * if the result format changes breaking backwards-compatibility,
+ * e.g. serialization format.
+ */
+export const currentResultSchemaVersion: SchemaVersion = "v1"
+
+export const kubernetesCacheEntrySchema = runResultSchemaZod.extend({
+  namespaceStatus: namespaceStatusSchema.required(),
+})
+
+export type KubernetesCacheEntrySchema = typeof kubernetesCacheEntrySchema
+
+export function composeKubernetesCacheEntry({
+  result,
+  namespaceStatus,
+}: {
+  result: RunResult
+  namespaceStatus: NamespaceStatus
+}): z.input<KubernetesCacheEntrySchema> {
+  return {
+    ...result,
+    namespaceStatus,
+  }
+}
+
+export interface LoadResultParams<A extends CacheableAction> {
+  ctx: PluginContext
+  log: Log
+  action: A
+}
+
+export type ClearResultParams<A extends CacheableAction> = LoadResultParams<A>
+
+export interface StoreResultParams<A extends CacheableAction, R> {
+  ctx: PluginContext
+  log: Log
+  action: A
+  result: R
+}
+
+export class StructuredCacheKey {
+  private readonly projectName: string
+  private readonly actionKind: string
+  private readonly actionName: string
+  private readonly actionType: string
+  private readonly actionVersion: string
+
+  constructor({ ctx, action }: { ctx: PluginContext; action: CacheableAction }) {
+    this.projectName = ctx.projectName
+    this.actionKind = action.kind
+    this.actionName = action.name
+    this.actionType = action.type
+    this.actionVersion = action.versionString()
+  }
+
+  @Memoize()
+  public calculate(): string {
+    const key = `${this.projectName}--${this.actionType}.${this.actionName}--${this.actionVersion}`
+    const hash = hashSync(key, { algorithm: "sha1" })
+    return `${this.actionKind.toLowerCase()}-result--${hash.slice(0, 32)}`
```

Comment:
```suggestion
    const hash = hashSync(key, { algorithm: "sha256" })
    return `${this.actionKind.toLowerCase()}-result--${hash}`
```
should we maybe use an unshortened cryptographically strong hash algorithm to ensure we are guaranteed to be free of collisions? (Assuming we do the same for the action version)